src/interaction/bilevel_planning_explorer.py

```diff
@@ -0,0 +1,58 @@
+"""An explorer that uses bilevel planning with NSRTs."""
+
+from typing import Callable, List, Set, Tuple
+
+from gym.spaces import Box
+
+from predicators.src import utils
+from predicators.src.interaction.base_explorer import BaseExplorer
+from predicators.src.option_model import _OptionModelBase
+from predicators.src.planning import PlanningFailure, PlanningTimeout, \
+    sesame_plan
+from predicators.src.settings import CFG
+from predicators.src.structs import NSRT, Action, ParameterizedOption, \
+    Predicate, State, Task, Type
+
+
+class BilevelPlanningExplorer(BaseExplorer):
+    """BilevelPlanningExplorer implementation.
+
+    The approach is abstract: subclasses decide how to use the bilevel
+    planning _solve method to create an exploration strategy.
+    """
+
+    def __init__(self, predicates: Set[Predicate],
+                 options: Set[ParameterizedOption], types: Set[Type],
+                 action_space: Box, train_tasks: List[Task], nsrts: Set[NSRT],
+                 option_model: _OptionModelBase) -> None:
+
+        super().__init__(predicates, options, types, action_space, train_tasks)
+        self._nsrts = nsrts
+        self._option_model = option_model
+        self._num_calls = 0
+
+    def _solve(
+        self, task: Task, timeout: int
+    ) -> Tuple[Callable[[State], Action], Callable[[State], bool]]:
+
+        # Ensure random over successive calls.
+        seed = self._seed + self._num_calls
+        try:
```

Comment:
sort of weird to replicate all this here, why not just pass approach.solve into the init?

Reply:
see comment below